File: glscan/src/main/java/com/dming/smallScan/ScannerView.java
package com.dming.smallScan;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.RadialGradient;
import android.graphics.Rect;
import android.graphics.Shader;
import android.util.AttributeSet;
import android.view.View;


/**
 * Created by DMing
 */

public class ScannerView extends View {
    private static final int[] SCANNER_ALPHA = {0, 64, 128, 192, 255, 192, 128, 64};
    private static final long ANIMATION_DELAY = 90L;
    private Paint mBgPaint;
    private Paint mCornerPaint;
    private Paint mFocusFramePaint;
    private Paint mLaserPaint;

    private int mCornerLength;
    private int mCornerThick;
    private int mFocusLineThick;
    private int scannerAlpha;

    private Rect mScannerRect;
    private RadialGradient mRadialGradient;


    public ScannerView(Context context) {
        this(context, null);
    }

    public ScannerView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    /**
     * 一些初始化操作
     */
    private void init() {

        int tingColor = 0xFF00FF00;
        int bgColor = 0x33000000;

        mBgPaint = new Paint();
        mCornerPaint = new Paint();
        mFocusFramePaint = new Paint();
        mLaserPaint = new Paint();

        mBgPaint.setColor(bgColor);
        mBgPaint.setAntiAlias(true);

        mCornerPaint.setAntiAlias(true);
        mCornerPaint.setColor(tingColor);

        mFocusFramePaint.setAntiAlias(true);
        mFocusFramePaint.setColor(Color.WHITE);

        mLaserPaint.setAntiAlias(true);
        mLaserPaint.setColor(tingColor);

        mCornerLength = 100;
        mCornerThick = 15;
        mFocusLineThick = 1;
        scannerAlpha = 0;
    }

    public void setScannerRect(Rect scannerRect) {
        mScannerRect = scannerRect;
        mRadialGradient = new RadialGradient(mScannerRect.width() / 2,
                10, mScannerRect.width() / 2,
                Color.RED, Color.BLUE, Shader.TileMode.CLAMP);
        mLaserPaint.setShader(mRadialGradient);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (mScannerRect == null) return;
        int width = getWidth();
        int height = getHeight();
        canvas.drawRect(0, 0, width, mScannerRect.top, mBgPaint);
        canvas.drawRect(0, mScannerRect.top, mScannerRect.left, mScannerRect.bottom + 1, mBgPaint);
        canvas.drawRect(mScannerRect.right + 1, mScannerRect.top, width, mScannerRect.bottom + 1, mBgPaint);
        canvas.drawRect(0, mScannerRect.bottom + 1, width, height, mBgPaint);
        drawCorner(canvas, mScannerRect);
        drawFocusRect(canvas, mScannerRect);
        drawLaser(canvas, mScannerRect);
        //实现动画效果
        postInvalidateDelayed(ANIMATION_DELAY, mScannerRect.left, mScannerRect.top, mScannerRect.right, mScannerRect.bottom);
    }

    /**
     * 绘制矩形框的四个角
     *
     * @param canvas
     * @param rect
     */
    private void drawCorner(Canvas canvas, Rect rect) {
        if (rect == null) return;
        //绘制左上角
        canvas.drawRect(rect.left, rect.top, rect.left + mCornerLength, rect.top + mCornerThick, mCornerPaint);
        canvas.drawRect(rect.left, rect.top, rect.left + mCornerThick, rect.top + mCornerLength, mCornerPaint);

        //绘制左下角
        canvas.drawRect(rect.left, rect.bottom - mCornerThick, rect.left + mCornerLength, rect.bottom, mCornerPaint);
        canvas.drawRect(rect.left, rect.bottom - mCornerLength, rect.left + mCornerThick, rect.bottom, mCornerPaint);

        //绘制右上角
        canvas.drawRect(rect.right - mCornerLength, rect.top, rect.right, rect.top + mCornerThick, mCornerPaint);
        canvas.drawRect(rect.right - mCornerThick, rect.top, rect.right, rect.top + mCornerLength, mCornerPaint);

        //绘制右下角
        canvas.drawRect(rect.right - mCornerLength, rect.bottom - mCornerThick, rect.right, rect.bottom, mCornerPaint);
        canvas.drawRect(rect.right - mCornerThick, rect.bottom - mCornerLength, rect.right, rect.bottom, mCornerPaint);
    }

    /**
     * 绘制聚焦框
     */
    private void drawFocusRect(Canvas canvas, Rect rect) {
        canvas.drawRect(rect.left + mCornerLength, rect.top, rect.right - mCornerLength, rect.top + mFocusLineThick, mFocusFramePaint);
        canvas.drawRect(rect.right - mFocusLineThick, rect.top + mCornerLength, rect.right, rect.bottom - mCornerLength, mFocusFramePaint);
        canvas.drawRect(rect.left + mCornerLength, rect.bottom - mFocusLineThick, rect.right - mCornerLength, rect.bottom, mFocusFramePaint);
        canvas.drawRect(rect.left, rect.top + mCornerLength, rect.left + mFocusLineThick, rect.bottom - mCornerLength, mFocusFramePaint);
    }

    /**
     * 绘制激光线
     */
    private void drawLaser(Canvas canvas, Rect rect) {
        mLaserPaint.setAlpha(SCANNER_ALPHA[scannerAlpha]);
        scannerAlpha = (scannerAlpha + 1) % SCANNER_ALPHA.length;
        int middle = rect.height() / 2 + rect.top;
        canvas.drawRect(rect.left, middle -5, rect.right, middle + 5, mLaserPaint);
    }


}

File: glscan/src/main/java/com/dming/smallScan/ScannerView.kt
package com.dming.smallScan

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View


/**
 * 扫描窗口的绘制view
 */
class ScannerView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) :
    View(context, attrs) {
    private lateinit var mBgPaint: Paint
    private lateinit var mCornerPaint: Paint
    private lateinit var mLineFramePaint: Paint
    private lateinit var mScanLinePaint: Paint

    private var mCornerSize: Int = 0
    private var mCornerThick: Int = 0
    private var mFrameLineWidth: Int = 0
    private var mScanLineSize: Int = 0
    private var mScanColor: Int = 0

    private var mScannerRect: Rect? = null
    private var mScannerLineRect: RectF? = null

    private var mAnimator: ValueAnimator? = null
    private var mAnimatorDuration: Long = 3000
    private var mAnimatedFraction: Float = 0f
    private var mScanLineMoveHeight: Float = 0f

    private var mScanLineDrawable: Drawable? = null
    private var mScanCornerDrawable: Drawable? = null

    /**
     * 一些初始化操作
     */
    fun initWithAttribute(gLViewParameter: GLViewParameter) {
        mScanLineDrawable = gLViewParameter.scanLine
        mScanCornerDrawable = gLViewParameter.scanCorner
        if (mScanCornerDrawable == null) { // 有扫描框图片用图片
            //角长度
            mCornerSize = gLViewParameter.scanCornerSize.toInt()
            //角宽
            mCornerThick = gLViewParameter.scanCornerThick.toInt()
        }
        // 扫描线尺寸
        mScanLineSize = gLViewParameter.scanLineWidth.toInt()
        // 扫描角和扫描框颜色
        mScanColor = gLViewParameter.scanColor
        // 框线宽
        mFrameLineWidth = gLViewParameter.scanFrameLineWidth.toInt()
        // 背景色和框线
        val scanBackgroundColor = gLViewParameter.scanBackgroundColor
        val scanFrameLineColor = gLViewParameter.scanFrameLineColor

        mBgPaint = Paint()
        mCornerPaint = Paint()
        mLineFramePaint = Paint()
        mScanLinePaint = Paint()

        mBgPaint.color = scanBackgroundColor
        mBgPaint.isAntiAlias = true

        val cornerColor = mScanColor and 0x00FFFFFF or 0xFF000000.toInt()
        mCornerPaint.isAntiAlias = true
        mCornerPaint.color = cornerColor

        mLineFramePaint.isAntiAlias = true
        mLineFramePaint.color = scanFrameLineColor

        mScanLinePaint.isAntiAlias = true

    }

    fun changeScanConfigure(scannerRect: Rect) {
        mScannerRect = scannerRect
        mScannerRect?.let { rect ->
            val ovalRect = RectF(
                0f,
                0f,
                scannerRect.width().toFloat(),
                mScanLineSize.toFloat()
            )
            val statColor = mScanColor and 0x00FFFFFF or 0xAA000000.toInt()
            val endColor = mScanColor and 0x00FFFFFF or 0x10000000
            val linearGradient = LinearGradient(
                ovalRect.left,
                ovalRect.bottom,
                ovalRect.width() / 2,
                ovalRect.bottom,
                endColor, statColor, Shader.TileMode.MIRROR
            )
            mScannerLineRect = ovalRect
            mScanLinePaint.shader = linearGradient
            if (mAnimator != null) {
                mAnimator!!.removeAllUpdateListeners()
            }
            if (ovalRect.height() > 0) {
                mAnimator = ValueAnimator.ofFloat(0f, rect.height().toFloat())
            }
            mAnimator?.let { animator ->
                animator.addUpdateListener { animation: ValueAnimator ->
                    mAnimatedFraction = animation.animatedFraction
                    postInvalidateOnAnimation(
                        rect.left,
                        rect.top,
                        rect.right,
                        rect.bottom
                    )
                }
                animator.cancel()
                animator.duration = mAnimatorDuration
                animator.repeatMode = ValueAnimator.RESTART
                animator.repeatCount = ValueAnimator.INFINITE
                animator.start()
            }
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        mAnimator?.cancel()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawBg(canvas) // 背景
        drawFrameLineRect(canvas) // 框边
        //
        if (mScanCornerDrawable != null) { // 四个角
            drawCornerDrawable(canvas)
        } else {
            drawCorner(canvas)
        }
        if (mScanLineDrawable != null) {  // 扫描线
            drawScanLineDrawable(canvas)
        } else {
            drawScanLine(canvas)
        }
    }

    private fun drawBg(canvas: Canvas) {
        mScannerRect?.let {
            canvas.drawRect(
                0f,
                0f,
                width.toFloat(),
                it.top.toFloat(),
                mBgPaint
            )
            canvas.drawRect(
                0f,
                it.top.toFloat(),
                it.left.toFloat(),
                it.bottom.toFloat(),
                mBgPaint
            )
            canvas.drawRect(
                it.right.toFloat(),
                it.top.toFloat(),
                width.toFloat(),
                it.bottom.toFloat(),
                mBgPaint
            )
            canvas.drawRect(
                0f,
                it.bottom.toFloat(),
                width.toFloat(),
                height.toFloat(),
                mBgPaint
            )
        }
    }

    /**
     * 绘制矩形框的四个角
     */
    private fun drawCorner(canvas: Canvas) {
        mScannerRect?.let {
            //绘制左上角
            canvas.drawRect(
                it.left.toFloat(),
                it.top.toFloat(),
                (it.left + mCornerSize).toFloat(),
                (it.top + mCornerThick).toFloat(),
                mCornerPaint
            )
            canvas.drawRect(
                it.left.toFloat(),
                it.top.toFloat(),
                (it.left + mCornerThick).toFloat(),
                (it.top + mCornerSize).toFloat(),
                mCornerPaint
            )

            //绘制左下角
            canvas.drawRect(
                it.left.toFloat(),
                (it.bottom - mCornerThick).toFloat(),
                (it.left + mCornerSize).toFloat(),
                it.bottom.toFloat(),
                mCornerPaint
            )
            canvas.drawRect(
                it.left.toFloat(),
                (it.bottom - mCornerSize).toFloat(),
                (it.left + mCornerThick).toFloat(),
                it.bottom.toFloat(),
                mCornerPaint
            )

            //绘制右上角
            canvas.drawRect(
                (it.right - mCornerSize).toFloat(),
                it.top.toFloat(),
                it.right.toFloat(),
                (it.top + mCornerThick).toFloat(),
                mCornerPaint
            )
            canvas.drawRect(
                (it.right - mCornerThick).toFloat(),
                it.top.toFloat(),
                it.right.toFloat(),
                (it.top + mCornerSize).toFloat(),
                mCornerPaint
            )

            //绘制右下角
            canvas.drawRect(
                (it.right - mCornerSize).toFloat(),
                (it.bottom - mCornerThick).toFloat(),
                it.right.toFloat(),
                it.bottom.toFloat(),
                mCornerPaint
            )
            canvas.drawRect(
                (it.right - mCornerThick).toFloat(),
                (it.bottom - mCornerSize).toFloat(),
                it.right.toFloat(),
                it.bottom.toFloat(),
                mCornerPaint
            )
        }
    }

    /**
     * 绘制矩形框的四个角
     */
    private fun drawCornerDrawable(canvas: Canvas) {
        mScanCornerDrawable?.let { scanCornerDrawable ->
            mScannerRect?.let { rect ->
                scanCornerDrawable.setBounds(
                    rect.left,
                    rect.top,
                    rect.right,
                    rect.bottom
                )
            }
            scanCornerDrawable.draw(canvas)
        }
    }

    /**
     * 绘制框
     */
    private fun drawFrameLineRect(canvas: Canvas) {
        mScannerRect?.let {
            canvas.drawRect(
                (it.left + mCornerSize).toFloat(),
                it.top.toFloat(),
                (it.right - mCornerSize).toFloat(),
                (it.top + mFrameLineWidth).toFloat(),
                mLineFramePaint
            )
            canvas.drawRect(
                (it.right - mFrameLineWidth).toFloat(),
                (it.top + mCornerSize).toFloat(),
                it.right.toFloat(),
                (it.bottom - mCornerSize).toFloat(),
                mLineFramePaint
            )
            canvas.drawRect(
                (it.left + mCornerSize).toFloat(),
                (it.bottom - mFrameLineWidth).toFloat(),
                (it.right - mCornerSize).toFloat(),
                it.bottom.toFloat(),
                mLineFramePaint
            )
            canvas.drawRect(
                it.left.toFloat(),
                (it.top + mCornerSize).toFloat(),
                (it.left + mFrameLineWidth).toFloat(),
                (it.bottom - mCornerSize).toFloat(),
                mLineFramePaint
            )
        }
    }

    /**
     * 绘制扫描线
     */
    private fun drawScanLine(canvas: Canvas) {
        mScannerLineRect?.let { lineRect ->
            mScannerRect?.let {
                canvas.save()
                mScanLineMoveHeight = it.height() - lineRect.height()
                canvas.translate(
                    it.left.toFloat(), it.top + mScanLineMoveHeight * mAnimatedFraction
                )
                canvas.drawOval(lineRect, mScanLinePaint)
                canvas.restore()
            }
        }
    }

    /**
     * 绘制扫描线 Drawable
     */
    private fun drawScanLineDrawable(canvas: Canvas) {
        mScannerLineRect?.let { lineRect ->
            mScanLineDrawable?.let { scanLineDrawable ->
                mScannerRect?.let {
                    canvas.save()
                    mScanLineMoveHeight = it.height() - lineRect.height()
                    canvas.translate(
                        it.left.toFloat(), it.top + mScanLineMoveHeight * mAnimatedFraction
                    )
                    scanLineDrawable.setBounds(
                        lineRect.left.toInt(),
                        lineRect.top.toInt(),
                        lineRect.right.toInt(),
                        lineRect.bottom.toInt()
                    )
                    scanLineDrawable.draw(canvas)
                    canvas.restore()
                }
            }
        }
    }

}